### DISCUSSION TITLE
Monkeypatching a state machine

### DISCUSSION THREAD

**Discussion by HerbCaudill** [statelyai/xstate#4787] [Q&A] [answered]
With the [@localfirst/auth connection protocol](https://github.com/local-first-web/auth/blob/main/packages/auth/src/connection/Connection.ts), I'd like to be able to test against various possible attacks by surgically modifying the state machine to simulate a misbehaving or adversarial client. 

I'd normally think of doing this with dependency injection - specifically by passing the actions and guards, as well as the machine definition itself, into the class that houses the machine. But the `setup().createMachine()` API really really wants everything to be defined inline, and in my experience so far even resists extracting any part of it to local variables, for example. 

A broad question is, is there another way to achieve what I want that fits more naturally with XState's model?

A more focused question is, if I were to pursue the dependency injection approach, can you point me to an example of instantiating a machine with manually typed, stand-alone actions/guards/config/etc.?

**Comment by davidkpiano** [comment, marked as answer]
After you setup the state machine using `setup().createMachine()`, you can still have a machine with provided implementations:

```ts
const modifiedMachine = machine.provide({
  // ...
})
```

Would that help for your use-case?

**Reply by HerbCaudill** [reply]
Ah - that should do the trick, thanks!
